src/types/mod.rs

```diff
@@ -774,3 +808,171 @@ impl<T: Object> ValueInternal for JsFunction<T> {
+        args.append(&mut self.args);
+        self
+    }
+
+    /// Make the function call. If the function returns without throwing, the result value
+    /// is downcast to the type `V`, throwing a `TypeError` if the downcast fails.
+    pub fn call<'b, C: Context<'b>, V: Value>(self, cx: &mut C) -> JsResult<'b, V> {
+        let v = self.callee.call(cx, self.this, self.args)?;
+        v.downcast_or_throw(cx)
+    }
+
+    /// Make the function call for side effect, discarding the result value. This method is
+    /// preferable to [`call()`](crate::types::FunctionCall::call) when the result value is
+    /// not needed, since it does not require specifying a result type.
+    pub fn exec<'b, C: Context<'b>>(self, cx: &mut C) -> NeonResult<()> {
+        let _ = self.callee.call(cx, self.this, self.args)?;
+        Ok(())
+    }
+}
+
+impl<'a> Call<'a> {
+    /// Set the value of `this` for the function call. Once a call has a `this` binding
+    /// specified, it is required to be a [`FunctionCall`](crate::types::FunctionCall).
+    pub fn this<V: Value>(self, this: Handle<'a, V>) -> FunctionCall<'a> {
+        FunctionCall {
+            callee: self.callee,
+            this: this.upcast(),
+            args: self.args,
+        }
+    }
+
+    /// Add an argument to the arguments list.
+    pub fn arg<V: Value>(mut self, arg: Handle<'a, V>) -> Self {
+        self.args.push(arg.upcast());
+        self
+    }
+
+    /// Add multiple arguments to the arguments list.
+    pub fn args<A: Arguments<'a>>(mut self, args: A) -> Self {
+        args.append(&mut self.args);
+        self
+    }
+
+    /// Call the function as a constructor (like a JavaScript `new` expression).
+    /// If the function returns without throwing, returns the resulting object.
+    pub fn new<'b, C: Context<'b>>(self, cx: &mut C) -> JsResult<'b, JsObject> {
+        self.callee.construct(cx, self.args)
+    }
+
+    /// Make the function call. If the function returns without throwing, the result value
+    /// is downcast to the type `V`, throwing a `TypeError` if the downcast fails.
+    pub fn call<'b, C: Context<'b>, V: Value>(self, cx: &mut C) -> JsResult<'b, V> {
+        let undefined: Handle<JsValue> = cx.undefined().upcast();
+        let v = self.callee.call(cx, undefined, self.args)?;
+        v.downcast_or_throw(cx)
+    }
+
+    /// Make the function call for side effect, discarding the result value. This method is
+    /// preferable to [`call()`](crate::types::Call::call) when the result value is not
+    /// needed, since it does not require specifying a result type.
+    pub fn exec<'b, C: Context<'b>>(self, cx: &mut C) -> NeonResult<()> {
+        let undefined: Handle<JsValue> = cx.undefined().upcast();
+        let _ = self.callee.call(cx, undefined, self.args)?;
+        Ok(())
+    }
+}
+
+/// The trait for specifying arguments in a [`Call`](crate::types::Call) or
+/// [`FunctionCall`](crate::types::FunctionCall).
+pub trait Arguments<'a> {
```

Comment:
Do we want to allow users to implement this `trait`? It _could_ be nice to allow users to do it, but if we don't make this `Sealed`, I would like to bikeshed the shape a bit more.

Reply:
No I agree, it should be sealed at least for now. I thought I remembered you had done some sealed traits before but my greps aren't turning up anything -- what conventions do you like to use for sealed traits?